csharp/src/Microsoft.ML.OnnxRuntime/OrtFloat16.shared.cs

```diff
@@ -0,0 +1,1372 @@
+            singleBits += ((uint)bfloatBits & 1) + RoundingBase;
+            bfloatBits = BitOpsUtils.SingleBitsToBFloat16Bits(singleBits);
+            return new BFloat16(bfloatBits);
+        }
+
+        /// <summary>
+        /// Explicitly converts a BFloat16 value to its nearest representable <see cref="float" /> value.
+        /// </summary>
+        /// <param name="value">The value to convert.</param>
+        /// <returns><paramref name="value" /> converted to its nearest representable <see cref="float" /> value.</returns>
+        public static explicit operator float(BFloat16 value)
+        {
+            bool sign = IsNegative(value);
+            int exp = value.BiasedExponent;
+            uint sig = value.TrailingSignificand;
+
+            if (exp == MaxBiasedExponent)
+            {
+                if (sig != 0)
+                {
+                    // Shift sig left 54 bits to get a 64-bit integer
+                    // to cut off all but 8 bits of the significant
+                    return BitOpsUtils.CreateSingleNaN(sign, (ulong)sig << 56);
+                }
+                return sign ? float.NegativeInfinity : float.PositiveInfinity;
+            }
+
+            if (exp == 0 && sig == 0)
+            {
+                // Positive / Negative zero
+                return (sign) ? -0.0f : 0.0f;
+            }
+
+            // All subnormal numbers in BFloat16 would be also subnormal in FP32 because they
+            // share the exponent.
+            uint singleBits = BitOpsUtils.BFloat16BitsToSingleBits(value.value);
+            return BitOpsUtils.UInt32BitsToSingle(singleBits);
+        }
+
+        /// <summary>
+        /// Flips the sign. NaNs are not affected.
+        /// IEEE 754 specifies NaNs to be propagated
+        /// </summary>
+        /// <param name="value"></param>
+        /// <returns></returns>
+        public static BFloat16 Negate(BFloat16 value)
+        {
+            return IsNaN(value) ? value : new BFloat16((ushort)(value.value ^ SignMask));
+        }
+
+        /// <summary>
+        /// The function returns true if the value is either NaN or zero.
+        /// </summary>
+        /// <param name="value">instance of BFloat16</param>
+        /// <returns>true if NaN or zero.</returns>
+        public static bool IsNaNOrZero(BFloat16 value)
+        {
+            uint abs = StripSign(value);
+            return (abs == 0 || abs > PositiveInfinityBits);
+        }
+
+        #region Utilities
+
+        private static bool AreZero(BFloat16 left, BFloat16 right)
+        {
+            // IEEE defines that positive and negative zero are equal, this gives us a quick equality check
+            // for two values by or'ing the private bits together and stripping the sign. They are both zero,
+            // and therefore equivalent, if the resulting value is still zero.
+            return (ushort)((left.value | right.value) & ~SignMask) == 0;
```

Comment:
Do we need `(ushort)`? Everything in this expression is ushort already.

Reply:
According to this ushort(s) are converted to int for bitwise operators in C#.
[https://learn.microsoft.com/en-us/dotnet/csharp/language-reference/operators/bitwise-and-shift-operators](https://learn.microsoft.com/en-us/dotnet/csharp/language-reference/operators/bitwise-and-shift-operators)